pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,529 @@
+    subst = {pymbolic.var(f"d{i}"): pymbolic.var("d")[i] for i in
+            range(base_kernel.dim)}
+    conv = SympyToPymbolicMapper()
+
+    if expr_multiplier == 1:
+        # if there's no expr_multiplier, only multiply the densities
+        return [int_g.copy(densities=tuple(density*density_multiplier
+            for density in int_g.densities))]
+
+    for knl, density in zip(int_g.source_kernels, int_g.densities):
+        if expr_multiplier == 1:
+            new_knl = knl.get_base_kernel()
+        else:
+            new_expr = conv(knl.postprocess_at_source(base_kernel_expr, sym_d)
+                    * expr_multiplier)
+            new_expr = pymbolic.substitute(new_expr, subst)
+            new_knl = ExpressionKernel(knl.dim, new_expr,
+                knl.get_base_kernel().global_scaling_const,
+                knl.is_complex_valued)
+        result.append(int_g.copy(target_kernel=new_knl,
+            densities=(density*density_multiplier,),
+            source_kernels=(new_knl,)))
+    return result
+
+
+def convert_int_g_to_base(int_g: IntG, base_kernel: ExpressionKernel) \
+        -> ExpressionT:
+    """Converts an *IntG* to an expression with *IntG*s having the
+    base kernel *base_kernel*.
+    """
+    result = 0
+    for knl, density in zip(int_g.source_kernels, int_g.densities):
+        result += _convert_int_g_to_base(
+                int_g.copy(source_kernels=(knl,), densities=(density,)),
+                base_kernel)
+    return result
+
+
+def _convert_int_g_to_base(int_g: IntG, base_kernel: ExpressionKernel) \
+        -> ExpressionT:
+    """Converts an *IntG* with only one source kernel to an expression with *IntG*s
+    having the base kernel *base_kernel*.
+    """
+    target_kernel = int_g.target_kernel.replace_base_kernel(base_kernel)
+    dim = target_kernel.dim
+
+    result = 0
+
+    density, = int_g.densities
+    source_kernel, = int_g.source_kernels
+    deriv_relation = get_deriv_relation_kernel(source_kernel.get_base_kernel(),
+        base_kernel, hashable_kernel_arguments=(
+            hashable_kernel_args(int_g.kernel_arguments)))
+
+    const = deriv_relation[0]
+    # NOTE: we set a dofdesc here to force the evaluation of this integral
+    # on the source instead of the target when using automatic tagging
+    # see :meth:`pytential.symbolic.mappers.LocationTagger._default_dofdesc`
+    dd = pytential.sym.DOFDescriptor(None,
+            discr_stage=pytential.sym.QBX_SOURCE_STAGE1)
+    const *= pytential.sym.integral(dim, dim-1, density, dofdesc=dd)
+
+    if const != 0 and target_kernel != target_kernel.get_base_kernel():
+        # There might be some TargetPointMultipliers hanging around.
+        # FIXME: handle them instead of bailing out
+        return int_g
+
+    if source_kernel != source_kernel.get_base_kernel():
+        # We assume that any source transformation is a derivative
+        # and the constant when applied becomes zero.
+        const = 0
```

Comment:
Any way to check that? `assert isinstance(..., AxisSourceDerivative)` or something?

Reply:
Done in 2a598edd66681972228090aca82fdbef263f29b3